auth/scoperesolver.js

```diff
@@ -0,0 +1,338 @@
+  }
+
+  async reload() {
+    debug("Loading clients and roles");
+
+    // Load clients and roles in parallel
+    let clients = [];
+    let roles   = [];
+    await Promise.all([
+      // Load all clients on a simplified form:
+      // {clientId, accessToken, updateLastUsed}
+      // _computeFixedPoint() will construct the `_clientCache` object
+      this._Client.scan({}, {
+        handler: client => {
+          let lastUsedDate = new Date(client.details.lastDateUsed);
+          clients.push({
+            clientId:       client.clientId,
+            accessToken:    client.accessToken,
+            // Note that lastUsedDate should be updated, if it's out-dated by more
+            // than 6 hours (cheap way to know if it's been used recently)
+            updateLastUsed: lastUsedDate < taskcluster.fromNow('-6h')
+          });
+        }
+      }),
+      // Load all roles on a simplified form: {roleId, scopes}
+      // _computeFixedPoint() will later add the `expandedScopes` property
+      this._Role.scan({}, {
+        handler(role) {
+          let scopes = role.scopes;
+          if (!role.roleId.endsWith('*')) {
+            // Ensure identity, if role isn't a prefix pattern. Obviously,
+            // 'assume:ab' which matches 'assume:a*' doesn't have 'assume:a*'
+            // by the identity relation. But for non-prefix patterns, the
+            // identify relation implies that you have 'assume:<roleId>'.
+            // This speeds up fixed-point computation, and means that if you
+            // have a match without any *, then you can look up the role, and
+            // not have to worry about any prefix patterns that may also match
+            // as they are already saturated.
+            scopes = _.union(scopes, ['assume:' + role.roleId]);
+          }
+          roles.push({roleId: role.roleId, scopes});
+        }
+      })
+    ]);
+
+
+    // Set _roles and _clients at the same time and immediately call
+    // _computeFixedPoint, so anyone using the cache is using a consistent one.
+    this._roles = roles;
+    this._clients = clients;
+    this._computeFixedPoint();
+  }
+
+  /** Compute fixed point over this._roles, and construct _clientCache */
+  _computeFixedPoint() {
+    // Add initial value for expandedScopes for each role R
+    for (let role of this._roles) {
+      role.expandedScopes = _.clone(role.scopes);
+    }
+
+    // Compute fixed-point of roles.expandedScopes for each role R
+    for (let R of this._roles) {
+      let isFixed = false;
+      while (!isFixed) {
+        isFixed = true; // assume we have fixed point
+        for (let role of this._roles) {
+          // if R can assume role, then we union the scope sets, as there is a
+          // finite number of strings here this will reach a fixed-point
+          let test = (scope) => ScopeResolver.grantsRole(scope, role.roleId);
+          if (R.expandedScopes.some(test)) {
+            let count = R.expandedScopes.length;
```

Comment:
why not check this length before and after the inner loop, rather than at every iteration?